### PR TITLE
Development branch

### PR DESCRIPTION
**Pull request by PantherJohn**
https://github.com/shadowsocks/shadowsocks-libev/issues/2419

### CODE REVIEW

**Comment by madeye** [comment]
Please rebase to the master branch first.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -32,7 +32,9 @@ set(SS_SHARED_SOURCES
         utils.c
         jconf.c
         json.c
+        cache.c
         netutils.c
+        relay.c
```
Comment (madeye) [inline src/CMakeLists.txt]: Rename it to tcprelay.c.

File: src/relay.h [hunk 2]
```diff
@@ -20,46 +20,25 @@
  * <http://www.gnu.org/licenses/>.
  */
 
-#ifndef _SERVER_H
-#define _SERVER_H
+#ifndef _RELAY_H
```
Comment (madeye) [inline src/relay.h]: Rename to `_TCP_RELAY_H`

**Comment by madeye** [comment]
Have you tried running all unit tests in `test/test.sh`? I cannot get any test passed with your changes.

**Comment by PantherJohn** [comment]
@madeye It's caused by the string null-termination issue. 

https://github.com/eLitetInc/shadowsocks-libev/blob/8096db706dbdf48f6dcbf528c5716fda493bb9fb/src/shadowsocks.h#L187-L210

Actually I was wondering why `cork_ip_init()` was needed. Look, just say if we accepted a connection to "https://1.1.1.1," then does that mean we have to force our remote server to unhappily reinterpret the domain name as an IP address? SO my suggestion here is to remove the lines above to **avoid potential issues in domain resolution** and **increase the efficiency in header parsing**.

**Comment by madeye** [comment]
It's way more faster than pass an IP address as hostname to remote server.

**Comment by PantherJohn** [comment]
@madeye fixed via a50b839

**Comment by PantherJohn** [comment]
@madeye fixed via a50b839

**Comment by madeye** [comment]
1. I still see all tests failing: all UDP forwarding timeout.
2. Remember to rebase the code.

**Comment by PantherJohn** [comment]
@madeye 
> Remember to rebase the code.

It’s barely possible to rebase the code. The master branch MUST be reverted to 34450e8d in order to resolve the merging conflict.

**Comment by madeye** [comment]
I think maybe the better solution is keeping these changes in your fork instead.

You changed too many fundamental things in this pull request, including the interfaces.

I'll pick some changes from your fork, if I find they are useful.

**Comment by PantherJohn** [comment]
@madeye That's basically why I'm asking for the `development` branch. In order for those `fundamental` changes to be settled, more people should be involved in the whole process of testing, debugging, and issue tracking.

**Comment by madeye** [comment]
Keep developing in this branch as you like.

And as mentioned in the very beginning, please don't expect everything you changed would get merged.

If you really want to open pull requests, make sure the change is small enough and there's no conflict.

**Pull request closed without merging**